src/server/hotspot_partition_calculator.cpp

```diff
@@ -0,0 +1,98 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "hotspot_partition_calculator.h"
+
+#include <algorithm>
+#include <math.h>
+#include <dsn/dist/fmt_logging.h>
+
+namespace pegasus {
+namespace server {
+
+void hotspot_partition_calculator::aggregate(const std::vector<row_data> &partitions)
+{
+    while (_app_data.size() > kMaxQueueSize - 1) {
+        _app_data.pop();
+    }
+    std::vector<hotspot_partition_data> temp(partitions.size());
+    for (int i = 0; i < partitions.size(); i++) {
+        temp[i] = std::move(hotspot_partition_data(partitions[i]));
+    }
+    _app_data.emplace(temp);
+}
+
+void hotspot_partition_calculator::init_perf_counter(const int perf_counter_count)
+{
+    std::string counter_name;
+    std::string counter_desc;
+    for (int i = 0; i < perf_counter_count; i++) {
+        string paritition_desc = _app_name + '.' + std::to_string(i);
+        counter_name = fmt::format("app.stat.hotspots@{}", paritition_desc);
+        counter_desc = fmt::format("statistic the hotspots of app {}", paritition_desc);
+        _points[i].init_app_counter(
+            "app.pegasus", counter_name.c_str(), COUNTER_TYPE_NUMBER, counter_desc.c_str());
+    }
+}
+
+void hotspot_partition_calculator::analysis(
+    const std::queue<std::vector<hotspot_partition_data>> &hotspot_app_data,
+    std::vector<::dsn::perf_counter_wrapper> &perf_counters)
+{
+    dassert(hotspot_app_data.back().size() == perf_counters.size(),
+            "partition counts error, please check");
+    std::vector<double> data_samples;
+    data_samples.reserve(hotspot_app_data.size() * perf_counters.size());
+    auto temp_data = hotspot_app_data;
+    double total = 0, sd = 0, avg = 0;
+    int sample_count = 0;
+    // avg: Average number
+    // sd: Standard deviation
+    // sample_count: Number of samples
+    while (!temp_data.empty()) {
+        for (auto partition_data : temp_data.front()) {
+            if (partition_data.total_qps - 1.00 > 0) {
+                data_samples.push_back(partition_data.total_qps);
+                total += partition_data.total_qps;
+                sample_count++;
+            }
+        }
+        temp_data.pop();
+    }
+    if (sample_count == 0) {
+        ddebug("hotspot_app_data size == 0");
+        return;
+    }
+    avg = total / sample_count;
+    for (auto data_sample : data_samples) {
+        sd += pow((data_sample - avg), 2);
```

Comment:
rename it to standard_deviation, so you can delete the comment above.
Same with `avg`

Reply:
This PR doesn't involve the refactor of the algorithm, which is contained in https://github.com/apache/incubator-pegasus/pull/592, thanks for staying tuned

Reply:
It's not the refactor of the algorithm, just modify some name of variables which are not very suitable.